### PR TITLE
Fix for ndarray_import when tensor imported is batch size 1.

### PR DESCRIPTION
**Pull request by galv**
Sorry for making a pullrequest instead of an issue, but it looks like pytorch does not always set the stride of its outermost dimension equal to 0 or the "expected stride" (i.e., the stride you would get if you increased the singleton dimension to a length greater than 1) when that dimension is singleton (i.e., that dimension's length is 1). Instead, it uses a stride of 1. `ndarray_import` assumes that singleton dimensions either have the "expected stride" or 0. If this isn't the case, you get an error about "incompatible function arguments".

I wrote a few reproducers for this issue, as well as a basic fix, though I admit it's a bit "ugly". However, since the only index that will ever be valid for a dimension of size 1 is 0, its stride will never be used anyway.

Edit: Upon further reflection, I don't believe this quick fix is entirely correct. What we want is to allow the slowest-changing dimension (So the first one in "C" ordering, or the last one in "F" ordering) to have any stride value when its length is equal to 1.

### CODE REVIEW

**Comment by wjakob** [comment]
The fix looks okay to me. Basically we don't care about strides for singleton or empty dimensions. While you are at it, could you change `(t.shape[i] == 1)` to  `t.shape[i] <= 1` (less than or equals, and removal of superfluous parenthesis.

Also, please rename the tests ("galvez") -- having your username in the will confuse people since it is not clear what it means. You could call them "pytorch_dim_1" and "pytorch_dim_2". Also add a comment (`see PR #162`).

**Comment by galv** [comment]
> The fix looks okay to me. Basically we don't care about strides for singleton or empty dimensions. While you are at it, could you change (t.shape[i] == 1) to t.shape[i] <= 1 (less than or equals, and removal of superfluous parenthesis.

You are right. I also further tested to make sure that the t.shape[i] == 0 case works for both numpy and pytorch (Notably, numpy sets all strides equal to 0 if t.shape[i] == 0 for any i).

> Also, please rename the tests ("galvez") -- having your username in the will confuse people since it is not clear what it means. You could call them "pytorch_dim_1" and "pytorch_dim_2". Also add a comment (see PR #162).

Yes, absolutely. Apologies for the initial sloppiness. I submitted the PR yesterday right before I had to head out for a previous commitment.

**Comment by galv** [comment]
I force pushed my updates. I removed an extra superfluous set of parentheses around `strides[i] == t.strides[i]` while doing this. I suppose I am not sure if you like those sorts of style changes being part of bug fix commits.

Let me know if you want anything else from me here. Thank you for your help.

**Comment by galv** [comment]
Let's hold off on merging this for a bit. I noticed some strange behaviors around shape[i] = 0 where i != 0 for pytorch. I will make a new commit explaining the issue and proposing a fix.

File: src/nb_ndarray.cpp
Comment (galv) [inline src/nb_ndarray.cpp:339]: This code is quite ugly. One nice thing would be to use a C++ iterator instead of lower-level indices. Just initialize the iterator to begin() if req_order = 'C', or rbegin() if req_order = 'F'. Also initialize end_it to either end() or rend().

Then we could drop the craziness going on at the bottom of the loop:

```
                    if ((req->req_order == 'C' && i == 0) ||
                        (req->req_order == 'F' && i == (uint32_t) (t.ndim - 1)))
                        break;
                    if (req->req_order == 'C') {
                        --i;
                    } else if (req->req_order == 'F') {
                        ++i;
                    }
```

and replace it with:

```
if (it == end_it) {
    break;
}
++it;
```

**Comment by galv** [comment]
Please read the commit message here: https://github.com/wjakob/nanobind/pull/162/commits/814aa4c191bde6bcfd688919844d23aa2e0b7901

I realized that the original code change as you suggested did not handle cases when shape[i] = 0 when i was not the slowest-changing dimension. Here is one suggested way of handling that appropriately, handling column-major order correctly. Note again that this rigamarole is basically due to pytorch's behavior regarding strides when an element of shape is equal to 0.. Numpy is perfectly fine as-is.

**Comment by galv** [comment]
Upon further reflection, I don't think https://github.com/wjakob/nanobind/commit/814aa4c191bde6bcfd688919844d23aa2e0b7901 is the "right" way to do things. I basically just need to encode the rule that if `t.shape[i] == 0` for any value of i, then we should accept any values for t.strides, since the tensor has 0 elements anyway. That would be much more concise than the current commit. I can probably get around to this by Monday (but wouldn't be upset if you went ahead and implemented that yourself).

**Comment by wjakob** [comment]
>  I basically just need to encode the rule that if t.shape[i] == 0 for any value of i, then we should accept any values for t.strides, since the tensor has 0 elements anyway.

That sounds good to me. I will wait for your updated PR (have a few other things on my plate)

**Comment by wjakob** [comment]
FYI, The failing test is a flake related to `capfd` in `pytest`, and not related to nanobind, I reported it separately.

**Comment by galv** [comment]
Alright, I made my change. I think this is good to go (assuming CI passes).

File: src/nb_ndarray.cpp
Comment (wjakob) [inline src/nb_ndarray.cpp:338]: Even if the CI passes, this cannot be merged as is because `std::any_of` is part of the `algorithm` header that we (intentionally) do not import. You will need to write this test manually using plain C/C++.

File: src/nb_ndarray.cpp
Comment (wjakob) [inline src/nb_ndarray.cpp:350]: Rather than including this as part of the loop condition, I would make the entire loop conditional on `!size_is_zero`. I am not a fan of comments within expressions, could you move them to the top? 

Also, note that your comment states ` if the shape of this dimension is 1,` but the actual test is `<= 1` which includes zero. Since that case is already covered by the `size_is_zero` condition, you could change it to `== 1` to remove potential confusion caused by the overlap.

**Review by wjakob — CHANGES_REQUESTED** [review]
A few more comments

**Comment by wjakob** [comment]
I realized the changes myself (want to push out 1.0.1 soon).

**Comment by galv** [comment]
No problem. Thank you for the help reviewing.

**Pull request merged**
